### ISSUE TITLE
Bug: Running CLI with only one rule fails to match any file

### ISSUE THREAD

**Issue by Pouja** [eslint/eslint#19361]
### Environment

Node version: v20.18.1
npm version: v10.8.2
Local ESLint version: v9.18.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin 24.2.0

### What parser are you using?

@typescript-eslint/parser

### What did you do?

When running eslint strictly using the CLI command it will report that no files are matching. But using the legacy engine through `ESLINT_USE_FLAT_CONFIG` works.

The use case is that we have a large repository (> 1M LoC) where we want to update rules from warn to error one at a time. And we want to use the auto fix of eslint to fix only a single rule, instead of all fixable problems.

The command I am trying to run for flat config:

```bash
node_modules/.bin/eslint \
--plugin '@typescript-eslint' \
--parser '@typescript-eslint/parser' \
--parser-options '{"project":"packages/awesome-helpers/tsconfig.json"}' \
--rule '{"@typescript-eslint/consistent-type-exports": "error"}' \
--no-config-lookup \
--no-cache \
"packages/awesome-helpers/src/**/*.ts"
```

### What did you expect to happen?

```
/path/to/packages/awesome-helpers/src/index.ts
  1:1  error  All exports in the declaration are only used as types. Use `export type`  @typescript-eslint/consistent-type-exports

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.
```

### What actually happened?

```
Oops! Something went wrong! :(

ESLint: 9.18.0

You are linting "packages/awesome-helpers/src/**/*.ts", but all of the files matching the glob pattern "packages/awesome-helpers/src/**/*.ts" are ignored.

If you don't want to lint these files, remove the pattern "packages/awesome-helpers/src/**/*.ts" from the list of arguments passed to ESLint.

If you do want to lint these files, explicitly list one or more of the files from this glob that you'd like to lint to see more details about why they are ignored.

  * If the file is ignored because of a matching ignore pattern, check global ignores in your config file.
    https://eslint.org/docs/latest/use/configure/ignore

  * If the file is ignored because no matching configuration was supplied, check file patterns in your config file.
    https://eslint.org/docs/latest/use/configure/configuration-files#specifying-files-with-arbitrary-extensions

  * If the file is ignored because it is located outside of the base path, change the location of your config file to be in a parent directory.
```

### Link to Minimal Reproducible Example

https://github.com/Pouja/eslint-auto-fix-issue

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

_I really like that we have switched to flat configs, thanks for the awesome work_

**Comment by nzakas** [comment]
Thanks for the report. I can reproduce this.

So, the problem here is because you aren't using a configuration file and you're trying to lint a `.ts` file, which is not configured by default. ESLint only configures `.js`, `.cjs`, and `.mjs` by default, so all of the command-line options that take the place of a config file are only applied to those files and the one you passed on the command line doesn't match, so it's skipped. This is clearly an oversight in how everything works together.

I'm not sure what the right solution is for this. On the one hand, if `--no-config-lookup` is passed, we could assume that you intend for other options to be applied to just the files passed on the command line, but that would also mean you wouldn't be able to just pass a directory because we'd have no idea what you wanted to lint in there. 

Another approach would be to add back the `--ext` command line option that we initially removed with flag config. That would allow you to explicitly pass which files should match the other options on the command line.

@eslint/eslint-team thoughts?

**Comment by JoshuaKGoldberg** [comment]
This is a ton of config to specify in the CLI:

> ```plaintext
> --plugin '@typescript-eslint' \
> --parser '@typescript-eslint/parser' \
> --parser-options '{"project":"packages/awesome-helpers/tsconfig.json"}' \
> --rule '{"@typescript-eslint/consistent-type-exports": "error"}' \
> --no-config-lookup \
> --no-cache \
> ```

ESLint _could_ add more options to allow specifying more and more things on the CLI. I'm not a fan of that direction. The spread of useful-in-narrow-scenarios cases like this is unbounded. I worry about the codebase and docs needlessly increasing in complexity to support them.

Instead, how about making a custom config file, and specifying the rule in there? It could even be passed in as an env var:

```js
// eslint.one-rule.config.js
import eslint from "@eslint/js";
import tseslint from "typescript-eslint";

export default tseslint.config(
  eslint.configs.recommended,
  tseslint.configs.recommended,
  {
    languageOptions: {
      parserOptions: {
        project: "packages/awesome-helpers/tsconfig.json",
      },
    },
    rules: {
      [process.env.ONE_RULE]: "error",
    },
  }
);
```

```shell
ONE_RULE="typescript-eslint/consistent-type-exports" eslint --config eslint.one-rule.config.js
```

---

>  `.ts` file, which is not configured by default

Separately, this could be changed IMO - especially now that Node.js supports them by default. But it might only make sense if the default parser supported TS syntax. Which is a much bigger change... Filed #19373.

**Comment by nzakas** [comment]
@JoshuaKGoldberg I'm also not a fan of adding so many command-line rules (I actually regret the ones we did add), however, `--ext` was an option that we had before and removed because we thought it was needed with flat config. This use case seems to indicate that it is still needed.

@Pouja the old config system worked differently. It always applied everything to files passed in. The new config system specifies rules, etc., based on file patterns and has no default pattern for TypeScript files.

**Comment by ItsTarik** [comment]
@nzakas in my case eslint finds all files and igonres them all, any explanation ?
i run this cmd: ` "lint": "eslint \"./src/**/*.ts\" --fix --debug"`

<img width="1185" alt="Image" src="https://github.com/user-attachments/assets/b6c83bca-84b7-4633-9b0c-14700431e940" />

**Comment by nzakas** [comment]
@ItsTarik this is not related to the issue. The relevant link is right in the error message:

https://eslint.org/docs/latest/use/configure/configuration-files#specifying-files-with-arbitrary-extensions

**Comment by Pouja** [comment]
@nzakas you remarks then further emphasize that using ESLint either with a custom config like @JoshuaKGoldberg presented or using it programmatically is they way to go for my use case.

For my part all my questions are answered and solved.

Thank you @nzakas and @JoshuaKGoldberg for your time!

**Comment by mdjermanovic** [comment]
> I'm not sure what the right solution is for this. On the one hand, if `--no-config-lookup` is passed, we could assume that you intend for other options to be applied to just the files passed on the command line, but that would also mean you wouldn't be able to just pass a directory because we'd have no idea what you wanted to lint in there.
> 
> Another approach would be to add back the `--ext` command line option that we initially removed with flag config. That would allow you to explicitly pass which files should match the other options on the command line.

I'm in favor of adding the `--ext` CLI option back. Technically, `--ext .ts` could add a `{ files: ["**/*.ts"] }` config object to `overrideConfig` array in order to enable linting of .ts files.

https://github.com/eslint/eslint/blob/8bcd820f37f2361e4f7261a9876f52d21bd9de8f/lib/cli.js#L171-L174

**Comment by fasttime** [comment]
Sounds good to add back the `--ext` CLI option for this use case.

**Comment by mdjermanovic** [comment]
It looks like we've agreed to add the `--ext` CLI option back. I can work on this.

**Comment by mdjermanovic** [comment]
Working on this.
